### PR TITLE
Add `Block.aload` method and remove `@sync_compatible` from `Block.load`

### PR DESCRIPTION
**Pull request by desertaxle**
This PR improves type hinting on `Block.load` by removing the `@sync_compatible` decorator and using `@async_dispatch` instead.

I had to update `@async_dispatch` to handle class methods and update the dispatch behavior in tasks and flows to match the behavior provided by `@sync_compatible`.

Also, `@inject_client` might be a thorn in our side because it makes wrapped functions async. To handle that in this PR, I removed `@inject_client` from `Block.load` and moved the client retrieval inside the method. There's a weird split where sometimes we can use the sync client, and other times we have to use the async client, but we can clean that up once we remove `@async_dispatch`.

Related to https://github.com/PrefectHQ/prefect/issues/16292

### CODE REVIEW

**Review by zzstoatzz — APPROVED** [review]
lgtm!

File: src/prefect/client/utilities.py [hunk 1]
```diff
@@ -71,9 +71,23 @@ async def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
...
+@overload
+def inject_client(
+    fn: Callable[P, R],
+) -> Callable[P, R]:
+    ...
+
+
+def inject_client(
+    fn: Callable[P, Union[Coroutine[Any, Any, R], R]],
+) -> Callable[P, Union[Coroutine[Any, Any, R], R]]:
```
Comment (mjpieters) [inline src/prefect/client/utilities.py:90]: @desertaxle Why did you add this overload?

The implementation uses `await` on `fn()` so it **must** produce an awaitable, and nowhere in the prefect codebase is `@inject_client` used on a regular function.

Comment (desertaxle) [reply src/prefect/client/utilities.py:90]: Yep, you're absolutely right. I was working with `@inject_client` while I was making `Block.load` sync. This was from something that ultimately didn't work, and I forgot to remove it. I'll make a note to remove this, but feel free to remove it if you're in the neighborhood.

Comment (mjpieters) [reply src/prefect/client/utilities.py:90]: @desertaxle already on it 😁  See #16463.
